kuka_rox_hardware_interface/src/robot_manager_node.cpp

```diff
@@ -309,21 +319,112 @@ RobotManagerNode::on_deactivate(const rclcpp_lifecycle::State &)
     RCLCPP_ERROR(get_logger(), "Could not stop controllers");
     return ERROR;
   }
+  controller_handler_.ApproveControllerDeactivation();
 
   RCLCPP_INFO(get_logger(), "Successfully stopped controllers");
   return SUCCESS;
 }
 
 bool RobotManagerNode::onControlModeChangeRequest(int control_mode)
 {
-  if (control_mode_map_.find(control_mode) != control_mode_map_.end()) {
+  try {
+    RCLCPP_INFO(get_logger(), "Control mode change process has started");
+    if (control_mode == kroshu_ros2_core::ControllerHandler::CARTESIAN_POSITION_CONTROL_MODE ||
+      control_mode == kroshu_ros2_core::ControllerHandler::CARTESIAN_IMPEDANCE_CONTROL_MODE ||
+      control_mode == kroshu_ros2_core::ControllerHandler::WRENCH_CONTROL_MODE)
+    {
+      RCLCPP_ERROR(get_logger(), "Tried to change a not implemented control mode");
+      // TODO(Svastits): this can be removed if rollback is implemented properly
+      this->on_deactivate(get_current_state());
+      return false;
+    }
+
+    auto controller_request =
+      std::make_shared<SwitchController::Request>();
+    std::pair<std::vector<std::string>, std::vector<std::string>> switchControllers;
+
+    // Activate controllers
+    if (get_current_state().id() == lifecycle_msgs::msg::State::PRIMARY_STATE_ACTIVE) {
+      // The driver is in active state
+
+      // Asks for witch controller to activate and deactivate
+      switchControllers = controller_handler_.GetControllersForSwitch(
+        kroshu_ros2_core::ControllerHandler::ControlMode(
+          control_mode));
+
+      // Call request for activateing controllers for the new control mode
+      if (!switchControllers.first.empty()) {
+        controller_request->activate_controllers = switchControllers.first;
+        controller_request->strictness = SwitchController::Request::STRICT;
+        auto controller_response =
+          kroshu_ros2_core::sendRequest<SwitchController::Response>(
+          change_controller_state_client_, controller_request, 0, 2000
+          );
+        if (!controller_response || !controller_response->ok) {
+          RCLCPP_ERROR(get_logger(), "Could not activate controllers for new control mode");
+          // TODO(Svastits): this can be removed if rollback is implemented properly
+          this->on_deactivate(get_current_state());
+          return false;
+        }
+      }
+      controller_handler_.ApproveControllerActivation();
+    }
+
+    // Publish the control mode to controller handler
+    auto message = std_msgs::msg::UInt32();
+    message.data = control_mode;
+    control_mode_pub_->publish(message);
+
+    if (get_current_state().id() == lifecycle_msgs::msg::State::PRIMARY_STATE_ACTIVE) {
```

Comment:
why are we checking for active state twice?

Reply:
Sorry, I'm not sure I understand that correctly. Can the get_current_state().id() at line 351 return a different value than the one at line 383? If so, this might be okay, but if the answer is no, then the second check is redundant.

Reply:
modified: The current state check called only once